File: Day88/Part2/shrinking_dungeon.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>

using namespace std;

// Problem: Dungeon With Shrinking State Space
// Grid N x M.
// Every step, 1 random cell becomes forbidden.
// Max blocks B.
// Goal: Certainty of reaching Target.
// Strategy: Can we reach Target before it gets blocked (or path cut)?
// Condition: Shortest Path L. If L <= steps_until_impossible?
// If B is small, we just need L <= safe_limit.
// But blocks are adversarial/random.
// If Target is blocked, game over.
// Can Target be blocked? "Satu sel acak menjadi terlarang".
// If Target is picked, we lose.
// Unless Target is immune? Usually Start/Target immune.
// Assume Start/Target immune.
// We need to bypass B blocked cells.
// Connectivity: If Node Connectivity (Start -> Target) > B, we can survive any B blocks?
// Not exactly. Time aspect: blocks happen sequentially.
// But if connectivity > B, there are > B disjoint paths.
// Even if they block optimally (worst case), they can only cut B paths.
// So if Connect > B, we win.
// This is rigorous.
// Max Flow with vertex capacities = 1.
// Source = Neighbors of Start. Sink = Neighbors of Target.
// Start/Target infinite capacity/immune.
// Intermediate nodes capacity 1.
// If MaxFlow > B, "Possible".
// Note: This assumes blocks are placed *anywhere* (worst case).
// "Random" with "Certainty" implies satisfying Worst Case.

struct Edge {
    int to;
    int capacity;
    int flow;
    int rev;
};

vector<vector<Edge>> adj;
vector<int> level;
vector<int> ptr;

void add_edge(int from, int to, int cap) {
    adj[from].push_back({to, cap, 0, (int)adj[to].size()});
    adj[to].push_back({from, 0, 0, (int)adj[from].size() - 1});
}

bool bfs(int s, int t) {
    fill(level.begin(), level.end(), -1);
    level[s] = 0;
    queue<int> q;
    q.push(s);
    while (!q.empty()) {
        int v = q.front();
        q.pop();
        for (const auto& edge : adj[v]) {
            if (edge.capacity - edge.flow > 0 && level[edge.to] == -1) {
                level[edge.to] = level[v] + 1;
                q.push(edge.to);
            }
        }
    }
    return level[t] != -1;
}

int dfs(int v, int t, int pushed) {
    if (pushed == 0) return 0;
    if (v == t) return pushed;
    for (int& cid = ptr[v]; cid < adj[v].size(); ++cid) {
        auto& edge = adj[v][cid];
        int tr = edge.to;
        if (level[v] + 1 != level[tr] || edge.capacity - edge.flow == 0) continue;
        int push = dfs(tr, t, min(pushed, edge.capacity - edge.flow));
        if (push == 0) continue;
        edge.flow += push;
        adj[tr][edge.rev].flow -= push;
        return push;
    }
    return 0;
}

int dinic(int s, int t) {
    int flow = 0;
    while (bfs(s, t)) {
        fill(ptr.begin(), ptr.end(), 0);
        while (int pushed = dfs(s, t, 1e9)) {
            flow += pushed;
        }
    }
    return flow;
}

int main() {
    int N, M, B;
    if (!(cin >> N >> M >> B)) return 0;

    int StartR, StartC, TargetR, TargetC;
    cin >> StartR >> StartC >> TargetR >> TargetC;

    // Node split for vertex capacity:
    // In-node: u, Out-node: u + NM
    // Capacity u -> u+NM is 1.
    // Neighbor u_out -> v_in is Inf.
    // Start and Target have Inf internal capacity (immune).

    int num_cells = N * M;
    int S = StartR * M + StartC;
    int T = TargetR * M + TargetC;

    // Source S_out, Sink T_in?
    // Actually standard vertex disjoint:
    // Split all nodes i into i_in and i_out. Edge i_in -> i_out cap 1.
    // Start and Target: cap Inf.
    // Adjacency u, v: u_out -> v_in cap Inf.
    
    int total_nodes = 2 * num_cells;
    adj.resize(total_nodes);
    level.resize(total_nodes);
    ptr.resize(total_nodes);

    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < M; ++c) {
            int u = r * M + c;
            int u_in = u;
            int u_out = u + num_cells;

            // Internal capacity
            if (u == S || u == T) {
                add_edge(u_in, u_out, 1e9);
            } else {
                add_edge(u_in, u_out, 1);
            }

            // Neighbors
            int dr[] = {0, 0, 1, -1};
            int dc[] = {1, -1, 0, 0};
            for (int k = 0; k < 4; ++k) {
                int nr = r + dr[k];
                int nc = c + dc[k];
                if (nr >= 0 && nr < N && nc >= 0 && nc < M) {
                    int v = nr * M + nc;
                    int v_in = v;
                    // u_out -> v_in
                    add_edge(u_out, v_in, 1e9);
                }
            }
        }
    }

    int max_paths = dinic(S, T + num_cells); // Flow from Start_in to Target_out

    if (max_paths > B) {
        cout << "Possible" << endl;
    } else {
        cout << "Impossible" << endl;
    }

    return 0;
}

File: Day88/Part2/shrinking_dungeon.py
import sys

# Max Flow (Dinic) implementation
class Dinic:
    def __init__(self, n):
        self.n = n
        self.graph = [[] for _ in range(n)]
        self.level = []

    def add_edge(self, u, v, capacity):
        self.graph[u].append([v, capacity, len(self.graph[v])])
        self.graph[v].append([u, 0, len(self.graph[u]) - 1])

    def bfs(self, s, t):
        self.level = [-1] * self.n
        self.level[s] = 0
        queue = [s]
        while queue:
            u = queue.pop(0)
            for v, cap, rev in self.graph[u]:
                if cap > 0 and self.level[v] < 0:
                    self.level[v] = self.level[u] + 1
                    queue.append(v)
        return self.level[t] >= 0

    def dfs(self, u, t, flow, ptr):
        if u == t or flow == 0:
            return flow
        for i in range(ptr[u], len(self.graph[u])):
            ptr[u] = i
            v, cap, rev = self.graph[u][i]
            if self.level[v] == self.level[u] + 1 and cap > 0:
                pushed = self.dfs(v, t, min(flow, cap), ptr)
                if pushed > 0:
                    self.graph[u][i][1] -= pushed
                    self.graph[v][rev][1] += pushed
                    return pushed
        return 0

    def max_flow(self, s, t):
        max_f = 0
        while self.bfs(s, t):
            ptr = [0] * self.n
            while True:
                pushed = self.dfs(s, t, float('inf'), ptr)
                if pushed == 0:
                    break
                max_f += pushed
        return max_f

def solve():
    try:
        line1 = sys.stdin.readline().split()
        if not line1: return
        N, M, B = map(int, line1)
        
        line2 = sys.stdin.readline().split()
        if not line2: return
        StartR, StartC, TargetR, TargetC = map(int, line2)
    except ValueError:
        return

    num_cells = N * M
    S = StartR * M + StartC
    T = TargetR * M + TargetC
    
    # Nodes: 0..num_cells-1 (In), num_cells..2*num_cells-1 (Out)
    dinic = Dinic(2 * num_cells)
    
    for r in range(N):
        for c in range(M):
            u = r * M + c
            u_in = u
            u_out = u + num_cells
            
            # Vertex Capacity
            cap = float('inf') if u == S or u == T else 1
            dinic.add_edge(u_in, u_out, cap)
            
            # Neighbors
            dr = [0, 0, 1, -1]
            dc = [1, -1, 0, 0]
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if 0 <= nr < N and 0 <= nc < M:
                    v = nr * M + nc
                    v_in = v
                    # Edge u_out -> v_in
                    dinic.add_edge(u_out, v_in, float('inf'))

    flow = dinic.max_flow(S, T + num_cells) # Actually flow from S_in (which goes to S_out) to T_out? 
    # Yes, S_in -> S_out (inf) -> neighbors_in.
    # T_in -> T_out (inf). Flow collected at T_out.
    
    if flow > B:
        print("Possible")
    else:
        print("Impossible")

if __name__ == "__main__":
    solve()
